### DISCUSSION TITLE
Pickle data was truncated

### DISCUSSION THREAD

**Discussion by With-Darkness** [KoljaB/RealtimeTTS#99] [Q&A]
I am happy to use your RealtimeTTS package. But I faced some errors and I couldn't fix it.
I want to use coquiEngine and execute coqui_test.py in the tests folder. 
I edited some code in this test as follows.

from RealtimeTTS import TextToAudioStream, CoquiEngine
import multiprocessing

if __name__ == '__main__':
    multiprocessing.freeze_support()  # Ensure proper multiprocessing initialization

    def dummy_generator():
        yield "Hey guys! These here are realtime spoken sentences based on local text synthesis. "
        yield "With a local, neuronal, cloned voice. So every spoken sentence sounds unique."
    
    try:
        engine = CoquiEngine(specific_model='v1.1')

        stream = TextToAudioStream(engine)

        print("Starting to play stream")
        stream.feed(dummy_generator()).play(log_synthesized_text=True)
    
        engine.shutdown()
    except Exception as e:
        print(e)

Unfortunately, there was an error that Pickle data was truncated.
Please help me.

**Comment by KoljaB** [comment]
This probably has to do with the usage of specific_model. Does the unchanged coqui_test.py work?

If you use specific_model but don't specify the local_models_path parameter it is expected that specific_model is a XTTS model that can be downloaded from huggingface.

So only "v2.0.0", "v2.0.1", "v2.0.2" and "v2.0.3" would be supported unless you use local_models_path parameter.

Greetings, 
Kolja

**Reply by With-Darkness** [reply]
> This probably has to do with the usage of specific_model. Does the unchanged coqui_test.py work?
> 
> If you use specific_model but don't specify the local_models_path parameter it is expected that specific_model is a XTTS model that can be downloaded from huggingface.
> 
> So only "v2.0.0", "v2.0.1", "v2.0.2" and "v2.0.3" would be supported unless you use local_models_path parameter.
> 
> Greetings,  Kolja

Thanks for your response. I did the original one first. But I couldn't see any other outputs. So I added some debug comments and a try-except statement. You can see my code as follows.

```
if __name__ == '__main__':
    from RealtimeTTS import TextToAudioStream, CoquiEngine

    def dummy_generator():
        yield "Hey guys! These here are realtime spoken sentences based on local text synthesis. "
        yield "With a local, neuronal, cloned voice. So every spoken sentence sounds unique."


    # for normal use with minimal logging:
    print("Preparing...")
    try:
        
        engine = CoquiEngine()
        print("engine setting completed.")
    
    
        # test with extended logging:
        # import logging
        # logging.basicConfig(level=logging.INFO)    
        # engine = CoquiEngine(level=logging.INFO)
    
    
        stream = TextToAudioStream(engine)
        
        print ("Starting to play stream")
        stream.feed(dummy_generator()).play(log_synthesized_text=True)
    
        engine.shutdown()
    except Exception as E:
        print(E)
```
But I can't see any other outputs only Preparing... message.
What's the reason?
![image](https://github.com/KoljaB/RealtimeTTS/assets/173931938/6ea2c4c6-7cc4-4097-8622-de77c17223f9)

Bonus Question: If I want to use other models such as tts_models/en/ljspeech/tacotron2-DDC?

**Comment by KoljaB** [comment]
> *But I can't see any other outputs only Preparing... message.*

Simplest possible reason: maybe you did not wait long enough. CoquiEngine needs a bit time to start.

If this does not work, please add maximum logging to see what exactly happens:

```python
if __name__ == '__main__':
    from RealtimeTTS import TextToAudioStream, CoquiEngine

    def dummy_generator():
        yield "Hey guys! These here are realtime spoken sentences based on local text synthesis. "
        yield "With a local, neuronal, cloned voice. So every spoken sentence sounds unique."

    import logging
    logging.basicConfig(level=logging.DEBUG)
    engine = CoquiEngine(level=logging.DEBUG)

    stream = TextToAudioStream(engine)
    
    print ("Starting to play stream")
    stream.feed(dummy_generator()).play(log_synthesized_text=True)

    engine.shutdown()
```

Please post the last lines of that output so I can see better, where CoquiEngine stops in initialization.

**Reply by With-Darkness** [reply]
> > _But I can't see any other outputs only Preparing... message._
> 
> Simplest possible reason: maybe you did not wait long enough. CoquiEngine needs a bit time to start.
> 
> If this does not work, please add maximum logging to see what exactly happens:
> 
> ```python
> if __name__ == '__main__':
>     from RealtimeTTS import TextToAudioStream, CoquiEngine
> 
>     def dummy_generator():
>         yield "Hey guys! These here are realtime spoken sentences based on local text synthesis. "
>         yield "With a local, neuronal, cloned voice. So every spoken sentence sounds unique."
> 
>     import logging
>     logging.basicConfig(level=logging.DEBUG)
>     engine = CoquiEngine(level=logging.DEBUG)
> 
>     stream = TextToAudioStream(engine)
>     
>     print ("Starting to play stream")
>     stream.feed(dummy_generator()).play(log_synthesized_text=True)
> 
>     engine.shutdown()
> ```
> 
> Please post the last lines of that output so I can see better, where CoquiEngine stops in initialization.

I did. And there is no additional log here. (I have been waiting more than 10 min)
![image](https://github.com/KoljaB/RealtimeTTS/assets/173931938/4967a8e2-517e-4478-bc88-71f8ca2d3b66)

**Comment by KoljaB** [comment]
Thanks for posting this. Something goes wrong within the coqui tts library this repo depends on. (The xtts load_checkpoint debug message is written directly before calling the [XTTS load_checkpoint method](https://github.com/idiap/coqui-ai-TTS/blob/dev/TTS/tts/models/xtts.py#L736) from Coqui TTS)

This is unfortunately a bit out of scope for me, because it happens in another library, which RealtimeTTS depends on. So I suggest:

Try some basic Coqui TTS examples first to verify coqui tts is installed and working on your system:
	https://github.com/idiap/coqui-ai-TTS/?tab=readme-ov-file#running-a-multi-speaker-and-multi-lingual-model

Then try an example using load_checkpoint:
	https://docs.coqui.ai/en/latest/models/xtts.html#id5

If the load_checkpoint example works, then RealtimeTTS should too. If it does not work, it's related to tts library and out of scope for me, you would need to log an issue in their github repo then.

**Reply by KoljaB** [reply]
Maybe it's worth trying to delete (or rename) the "v2.0.2" folder in the path you see in the last line of your screenshot. In case something went wrong during downloading of the model. Not sure if that helps though.

**Comment by With-Darkness** [comment]
My name is Joseph Dingess, known as With-Darkness on GitHub. I am eager to utilize your module. Following your recommendation, I have initiated a project. Specifically, I am interested in utilizing the English model rather than multilanguage models. Therefore, I have begun to follow the project accordingly.

import torch
from TTS.api import TTS
import logging

OUTPUT_PATH='audio1.wav'

# Get device
device = "cuda" if torch.cuda.is_available() else "cpu"

# Init TTS with the target model name
tts = TTS(model_name="tts_models/en/ljspeech/tacotron2-DDC", progress_bar=False).to(device)

# Run TTS
tts.tts_to_file(text="Hello Guys I am potter", file_path=OUTPUT_PATH)

It works well so tried this using your module as follows.
if __name__ == '__main__':
    from RealtimeTTS import TextToAudioStream, CoquiEngine

    def dummy_generator():
        yield "Hey guys! These here are realtime spoken sentences based on local text synthesis. "
        yield "With a local, neuronal, cloned voice. So every spoken sentence sounds unique."

    import logging
    logging.basicConfig(level=logging.DEBUG)
    engine = CoquiEngine(level=logging.DEBUG, model_name="tts_models/en/ljspeech/tacotron2-DDC", specific_model="")

    stream = TextToAudioStream(engine)

    print ("Starting to play stream")
    stream.feed(dummy_generator()).play(log_synthesized_text=True)

    engine.shutdown()

But there is still error like this.
[cid:edf5f883-5991-463f-8b65-9ec49238d504]
Please help me.  If you want to contact me out side of mail, you can find me in Discord josephdingess0926.
________________________________
